### ISSUE TITLE
New TUI spams lines when pulling multiple images causing you to lose your terminal history

### ISSUE THREAD

**Issue by clrxbl** [docker/compose#8800]
**Description**

@thaJeztah has asked me to open an issue after commenting on #8753 

> For your use-case, is there a specific reason why the new output is problematic for you? If so, could you describe your use-case? Perhaps there's enhancements to be made to address.

I'm not the person who created this issue but I have one complaint about the new output as well.
First of all, I love BuildKit's output in docker build. But, compose's implementation makes me question how this made it through an official 2.0 release.

https://user-images.githubusercontent.com/28601081/137222298-31ca5d3a-68be-4041-82e8-8d58cb8e8999.mp4

This can be reproduced in both Windows' command prompt, Windows Terminal and probably others. It literally spams thousands of lines to the terminal ruining the ability to scroll back. Sorry if I'm hijacking the issue but it seemed fairly fitting as I have to resort to rolling back to a 1.x release.

_Originally posted by @clrxbl in https://github.com/docker/compose/issues/8753#issuecomment-942773151_

**Steps to reproduce the issue:**
1. docker compose up on a docker-compose.yml file that contains multiple images & a small enough terminal window.

**Describe the results you received:**
See the above video

**Describe the results you expected:**
I should be able to scroll back and not lose all of my terminal history, even after the command is done.

**Additional information you deem important (e.g. issue happens only occasionally):**
It seems like this issue isn't present if you're reproducing it in a large enough terminal window (e.g. fullscreen)

**Output of `docker compose version`:**

```
Docker Compose version 2.0.1
```

**Output of `docker info`:**

```
WARNING: Plugin "/usr/local/lib/docker/cli-plugins/docker-buildx" is not valid: failed to fetch metadata: fork/exec /usr/local/lib/docker/cli-plugins/docker-buildx: no such file or directory
WARNING: Plugin "/usr/local/lib/docker/cli-plugins/docker-compose" is not valid: failed to fetch metadata: fork/exec /usr/local/lib/docker/cli-plugins/docker-compose: no such file or directory
WARNING: Plugin "/usr/local/lib/docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /usr/local/lib/docker/cli-plugins/docker-scan: no such file or directory
Client:
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
 Context:    default
 Debug Mode: false
 Plugins:

Server:
 Containers: 12
  Running: 0
  Paused: 0
  Stopped: 12
 Images: 11
 Server Version: 20.10.9
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8686ededfc90076914c5238eb96c883ea093a8ba.m
 runc version: v1.0.2-0-g52b36a2d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.60.1-microsoft-standard-WSL2
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.63GiB
 Name: DESKTOP-BQ26BOE-wsl
 ID: CWUC:IOEW:TCJZ:EZJS:RTO5:YSBV:X7BE:YHNS:MQNY:V3VQ:N3NI:I4T4
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: clrxbl
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details:**
Arch Linux WSL2 w/ systemd-genie

**Comment by ndeloof** [comment]
@crazy-max any thoughts?

**Comment by thaJeztah** [comment]
I don't think it just build, although multiple builds will definitely be adding to the output. The example video shows that pulling (if the compose-file has many images) can also be problematic if there's more pulls than fit on the screen.


Doing a quick test with a stack that uses a `postgres` image (which has various layers);

```yaml
services:
    foo:
        image: postgres
```

Produces this output (ignore the error at the end, it's just for illustration):


```bash
docker compose up
[+] Running 14/14
 ⠿ foo Pulled                                          14.6s
   ⠿ e5ae68f74026 Pull complete                         5.6s
   ⠿ 7b8fcc7e1ad0 Pull complete                         6.0s
   ⠿ 7527d03e2f77 Pull complete                         6.1s
   ⠿ 80e55689f4d0 Pull complete                         6.2s
   ⠿ 8a79eb6d69c9 Pull complete                         6.7s
   ⠿ 397705f2d093 Pull complete                         6.8s
   ⠿ de36ec4eb0a5 Pull complete                         6.9s
   ⠿ 08d878a022c1 Pull complete                         7.0s
   ⠿ 7677029670ff Pull complete                        11.4s
   ⠿ 1d24b3d9557e Pull complete                        11.5s
   ⠿ e085b018338c Pull complete                        11.5s
   ⠿ 063b09ff12e9 Pull complete                        11.6s
   ⠿ a39fee215a44 Pull complete                        11.7s
[+] Running 2/2
 ⠿ Network compose-progress_default  Created            0.1s
 ⠿ Container compose-progress-foo-1  Created            1.4s
Attaching to compose-progress-foo-1
compose-progress-foo-1  | Error: Database is uninitialized and superuser password is not specified.
compose-progress-foo-1  |        You must specify POSTGRES_PASSWORD to a non-empty value for the
compose-progress-foo-1  |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
compose-progress-foo-1  |
compose-progress-foo-1  |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
compose-progress-foo-1  |        connections without a password. This is *not* recommended.
compose-progress-foo-1  |
compose-progress-foo-1  |        See PostgreSQL documentation about "trust":
compose-progress-foo-1  |        https://www.postgresql.org/docs/current/auth-trust.html
compose-progress-foo-1 exited with code 1
```

While the pull is running, `Downloading` and `Extracting` progress bars are shown for each layer:

```
   ⠼ e5ae68f74026 Extracting      [==================================================>]  31.37MB/31.37MB    4.4s
   ⠼ 7b8fcc7e1ad0 Download complete                                                                         4.4s
   ⠼ 7527d03e2f77 Download complete                                                                         4.4s
   ⠼ 80e55689f4d0 Download complete                                                                         4.4s
   ⠼ 8a79eb6d69c9 Download complete                                                                         4.4s
   ⠼ 397705f2d093 Download complete                                                                         4.4s
   ⠼ de36ec4eb0a5 Download complete                                                                         4.4s
   ⠼ 08d878a022c1 Download complete                                                                         4.4s
   ⠼ 7677029670ff Downloading     [=============================>                     ]  53.55MB/91.23MB    4.4s
   ⠼ 1d24b3d9557e Download complete                                                                         4.4s
```


Based on that output, some ideas:

### Remove the `pull complete` lines after they have completed / after the image has fully downloaded

Collapse the section, and replace with a summary.

```
⠿ foo Pulled                                          14.6s
```

Perhaps the summary should have more details than just "pulled"; users may be interested to know _what image_ (and tag, digest?) was pulled, and how large it was in total (?), as well as the architecture that was pulled (?) haven't given that a lot of thinking yet;

```
⠿ foo Pulled postgres:latest (digest sha256:xxx) linux/amd64, 12345MB                  14.6s
```


### Use a single progress (or counter) per image during pull

Perhaps depending on the amount of available space or the number of services / pulls, perhaps it's possible to group the progress bars. The size would show the _total_ size / sum of all layers (I'm not sure if this works if layers are already downloaded before though);

```
postgres:latest  Downloading  3/13   [=============================>                     ]  53.55MB/91.23MB    4.4s
postgres:latest  Extracting   2/13   [==================================================>]  31.37MB/31.37MB    4.4s
```

**Comment by thaJeztah** [comment]
Looking at the output, these are also somewhat confusing:

```
[+] Running 14/14
[+] Running 2/2
```

Wondering if we should use different names for such sections (e.g. "Pulling images for services") or even if some actions should be grouped "per service". For example, the current output also shows:

```
[+] Running 3/3
 ⠿ Network compose-progress_default  Created                                             0.0s
 ⠿ Container compose-progress-bar-1  Created                                             3.2s
 ⠿ Container compose-progress-foo-1  Created                                             3.2s
```

Perhaps the `Container compose-progress-bar-1  Created` (etc.) should be under a `Service foo` group (not sure if "created" and "started" should be shown separately after it's done);

```bash
[+] Starting services 2/2
  [+] Service foo 3/3
    ⠿ Image postgres:latest pulled (digest sha256:xxx) linux/amd64, 12345MB                14.6s
    ⠿ Container compose-progress-foo-1  Created                                             3.2s
    ⠿ Container compose-progress-foo-1  Started                                             1.1s
  [+] Service bar 3/3
    ⠿ Image postgres:latest pulled (digest sha256:xxx) linux/amd64, 12345MB                14.6s
    ⠿ Container compose-progress-bar-1  Created                                             3.2s
    ⠿ Container compose-progress-bar-1  Started                                             1.1s
```

**Comment by thaJeztah** [comment]
While trying, I also noticed some "jittery" output if two services use the same image (this may be the same if they use different images, but layers that are shared between those images);

```yaml
services:
    foo:
        image: postgres
    bar:
        image: postgres
```

During the pull, the output is "jumpy". Perhaps because both `foo` and `bar` services try to show progress for the same layers:

```bash
docker compose down -v
docker rmi postgres

docker compose up
[+] Running 8/15
 ⠋ foo Pulling                                                                                                                                                                                       12.0s
   ⠿ 8a79eb6d69c9 Pull complete                                                                                                                                                                       4.9s
   ⠿ 397705f2d093 Pull complete                                                                                                                                                                       5.1s
   ⠿ de36ec4eb0a5 Pull complete                                                                                                                                                                       5.2s
   ⠏ 1d24b3d9557e Download complete                                                                                                                                                                   8.9s
   ⠏ a39fee215a44 Download complete                                                                                                                                                                   8.9s
 ⠋ bar Pulling                                                                                                                                                                                       12.0s
   ⠿ e5ae68f74026 Pull complete                                                                                                                                                                       3.4s
   ⠿ 7b8fcc7e1ad0 Pull complete                                                                                                                                                                       3.9s
   ⠿ 7527d03e2f77 Pull complete                                                                                                                                                                       4.0s
   ⠿ 80e55689f4d0 Pull complete                                                                                                                                                                       4.2s
   ⠿ 08d878a022c1 Pull complete                                                                                                                                                                       5.3s
   ⠏ 7677029670ff Extracting      [=============================================>     ]  83.56MB/91.23MB                                                                                              8.9s
   ⠏ e085b018338c Download complete                                                                                                                                                                   8.9s
   ⠏ 063b09ff12e9 Download complete                                                                                                                                                                   8.9s
```

After the pull completed, the output is a bit confusing, because _some_ layers are shown under service `foo`, and _some_ under `bar`

```bash
docker compose up
[+] Running 15/15
 ⠿ foo Pulled                                                                                                                                                                                        14.5s
   ⠿ 8a79eb6d69c9 Pull complete                                                                                                                                                                       4.9s
   ⠿ 397705f2d093 Pull complete                                                                                                                                                                       5.1s
   ⠿ de36ec4eb0a5 Pull complete                                                                                                                                                                       5.2s
   ⠿ e085b018338c Pull complete                                                                                                                                                                      11.3s
 ⠿ bar Pulled                                                                                                                                                                                        14.5s
   ⠿ e5ae68f74026 Pull complete                                                                                                                                                                       3.4s
   ⠿ 7b8fcc7e1ad0 Pull complete                                                                                                                                                                       3.9s
   ⠿ 7527d03e2f77 Pull complete                                                                                                                                                                       4.0s
   ⠿ 80e55689f4d0 Pull complete                                                                                                                                                                       4.2s
   ⠿ 08d878a022c1 Pull complete                                                                                                                                                                       5.3s
   ⠿ 7677029670ff Pull complete                                                                                                                                                                      11.2s
   ⠿ 1d24b3d9557e Pull complete                                                                                                                                                                      11.2s
   ⠿ 063b09ff12e9 Pull complete                                                                                                                                                                      11.4s
   ⠿ a39fee215a44 Pull complete                                                                                                                                                                      11.4s
[+] Running 3/3
 ⠿ Network compose-progress_default  Created                                                                                                                                                          0.0s
 ⠿ Container compose-progress-bar-1  Created                                                                                                                                                          3.2s
 ⠿ Container compose-progress-foo-1  Created                                                                                                                                                          3.2s
Attaching to compose-progress-bar-1, compose-progress-foo-1
```

Not sure what the solution to that would be;

- for the "two services use the _same image_", I guess compose could detect this case, and only do a pull once, but of course that won't help if the images are different (but share common layers).
- alternatively, do the "reverse", and make sure that layer progress is always shown under "both" (duplicate the progress in case it's shared); possibly challenging to do this

Hiding progress of individual layers (as mentioned in my previous comment) may help here as well.

**Comment by ericslandry** [comment]
@thaJeztah I can confirm that this is not just happening with `build`; I've been seeing this with `up` too in Compose 2.1.1 in Ubuntu 20.

I have no idea about the implementation, but could the live progress output be trimmed to the available space with a "snipped" indicator?

**Comment by ndeloof** [comment]
@thaJeztah buildkit indeed consider the build requests we schedule in parallel for services to be fully independent, and as such won't do any magic on concurrent pulls for same layers.

@ericslandry for your information, this output is fully managed by buildx.

**Comment by thaJeztah** [comment]
> this output is fully managed by buildx.

In my example, these are `pulls`, no build used, or do you mean it's using the buildx library code for this?

**Comment by ndeloof** [comment]
Maybe I'm wrong, I thought progress writer introduced by https://github.com/docker/compose-cli/pull/233 was a copy/paste from buildkit. Maybe a full re-implementation with the same UX in mind?
Does `docker buildx bake` have the same issue rendering build/pull events ?

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/buildkit/issues/2511

**Comment by ericslandry** [comment]
@ndeloof :  does buildx also control the part when containers are creating/starting?
```
[+] Running 43/43redacted-1                Started                 4.5s
 ⠿ Network redacted-default                 Created                 0.0s
 ⠿ Volume "redacted_a-data"                 Created                 0.0s
 ⠿ Volume "redacted_b-data"                 Created                 0.0s
 ⠿ Container redacted-a-1                   Started                 3.0s
 ⠿ Container redacted-b-1                   Started                 3.1s
 ⠿ Container redacted-c-1                   Started                 3.3s
 ⠿ Container redacted-postgres-1            Started                 3.3s
 ⠿ Container redacted-kafka-1               Started                10.0s
```

In a short windows (say 4 lines high), the problem still occurs, even without the pulls

**Comment by thaJeztah** [comment]
.

**Comment by BBaoVanC** [comment]
Was this fixed by #9476 (in release v2.6.0)?

**Comment by ericslandry** [comment]
Here's my use case to reproduce the problem.

- Docker 20.10.17
- Docker Compose v2.6.0
- This `compose.yaml`:
```yaml
x-service: &default-service
  image: nginx:latest
  command: >
    bash -c "sleep 10 && nginx -g 'daemon off;'"
  healthcheck:
    test: service nginx status || exit 1
    interval: 2s
    timeout: 1s
    retries: 200
    start_period: 600s

services:
  svc_a:
    <<: *default-service
  svc_b:
    <<: *default-service
    depends_on:
      svc_a:
        condition: service_healthy
  svc_c:
    <<: *default-service
    depends_on:
      svc_b:
        condition: service_healthy
  svc_d:
    <<: *default-service
    depends_on:
      svc_c:
        condition: service_healthy
  svc_e:
    <<: *default-service
    depends_on:
      svc_d:
        condition: service_healthy
  svc_f:
    <<: *default-service
    depends_on:
      svc_e:
        condition: service_healthy
  svc_g:
    <<: *default-service
    depends_on:
      svc_f:
        condition: service_healthy
  svc_h:
    <<: *default-service
    depends_on:
      svc_g:
        condition: service_healthy
  svc_i:
    <<: *default-service
    depends_on:
      svc_h:
        condition: service_healthy
  svc_j:
    <<: *default-service
    depends_on:
      svc_i:
        condition: service_healthy
  svc_k:
    <<: *default-service
    depends_on:
      svc_j:
        condition: service_healthy
  svc_l:
    <<: *default-service
    depends_on:
      svc_k:
        condition: service_healthy
```
- Make terminal 11 lines high
- `docker compose up -d`
- Notice terminal scroll

**Comment by storm1er** [comment]
@BBaoVanC The problem is still present if your docker-compose contains multiple services using different images

There's should be a toggle option (cli + env. var. ?) so the output is not a mess on complexe stack pull/build :(

**Comment by BBaoVanC** [comment]
@ericslandry I can reproduce using your config.

I think the reason that v2.6.0 fixed it for me was because it made it so only one service would show the pull progress at the same time. After that change, my terminal was never small enough to not be able to fit it all on one screen without scrolling.

Your config works fine for me if my terminal is tall enough to fit it all (about 23 lines), but if I do 11 lines like you said, I see the issue happen again.

**Comment by Codelica** [comment]
Boy is this a mess on a large compose pull.  How about an option to just suppress individual layer progress?  It doesn't take much to overshoot the line height of most terminals with a reasonable compose file.   Otherwise you have to either watch the fireworks, or resort to something like:

`docker compose ps --services |xargs -n 1 docker compose pull`

**Comment by michaelkrieger** [comment]
Have a large number of containers within a docker-compose file.
Same problem:
1. It spams the output buffer by jumping around like crazy
2. it flashes as fireworks like crazy while it's happening
3. The ultimate output simply says "[container] Pulled ... X.Xs" but provides no indications of which were up to date in the first place and which were actively pulled
4. The ultimately output is a mix of 'containername Pulled", "XXXX already exists" and "xxxx pull complete" with some items showing the breakout of parts and others not.

As pointed out, it's very easy to exceed a typical terminal height in this case.

Possible ideas:
1. Do the docker compose pull sequentially or in batches, which reduces the number of pulls at any given time and thus concludes some items before the next few are started
2. Spare the output of any that are "up to date" to the end.  A quick test of whether the image is up to date, and if it is, output nothing until after all are updated and then output the "XXXX is up to date"
3. A better terminal implementation

**Comment by bklaric** [comment]
A workaround that works for me is to add `--quiet-pull` to `docker compose up`. This doesn't show progress for individual images, only the service itself.

**Comment by ShadowLNC** [comment]
This is becoming significantly more problematic as our users get automatically opted in to Compose V2 - with the size of our compose file, `docker compose pull` is virtually unreadable.

I do believe it's the same as #9377 but the height required exceeds that of my built in monitor on my laptop, so it's difficult to verify.

**Comment by ZuluWhiskey** [comment]
I believe I have been having this issue.
Connecting to a common server, I was not getting the issue when on my mobile client (Termius for iOS) but was getting it on my desktop using SecureCRT where it was producing thousands of lines when pulling images.
I've just played with the settings on SecureCRT and found that if I disabled the "Line wrap" emulation mode, it stopped the issue from happening instantly.
![image](https://user-images.githubusercontent.com/35011199/193297358-c377b8b1-5755-4d75-94ee-7b77145cc554.png)

Apologies if this comment isn't of any relevance or significance but I found this issue when trying to troubleshoot the problem and wanted to share the fix that worked for me.

**Comment by ericslandry** [comment]
@laurazard Thank you very much for improving the situation! Unfortunately, my problem persists as per your PR comment:

"Detect when the number of events we have to display is > than terminal height, and adjust the output to omit child events when that's the case. This isn't a perfect solution (**if the number of services > terminal height we will still run into issues**, and the shortened output isn't very explicit), but it only kicks in in cases where the output would otherwise be worse, so it's an improvement over the current situation."

Would it be ok to leave this issue open or should a new issue be created?

**Comment by ndeloof** [comment]
Please open a fresh new issue.
The main question to address is "_how to display status for N services when N > terminal height?_".  Any suggestion is welcome

**Comment by thenick775** [comment]
I'm at least still seeing this on Docker Compose version `v2.17.3` with docker version `23.0.5`

**Comment by dcarbone** [comment]
I'm not sure line-height is the ultimate culprit here.  I'm seeing hectic-as-heck output even with single service build in a full height terminal window on a laptop with a built-in 1600x1200 display with scaling disabled.  It started happening with the release of v2.17.  With v2.16, output with the same services was as-expected.

**Comment by Mariusio** [comment]
> I'm at least still seeing this on Docker Compose version `v2.17.3` with docker version `23.0.5`

Same for me. I am using `Docker Desktop 4.19.0` on a M1 Macbook 16-inch, 2021 with Ventura 13.3.1.

Interestingly this bug only surfaced today after updating Docker Desktop from `4.17.x` to `4.19.0`, before the log output didn't jump and everything was smooth even with the same exact same environment and same `Dockerfile` and `docker-compose.yml`.

**Comment by williamdes** [comment]
Maybe all of you should comment on open issue #10256

**Comment by richardvrusso** [comment]
> I believe I have been having this issue. Connecting to a common server, I was not getting the issue when on my mobile client (Termius for iOS) but was getting it on my desktop using SecureCRT where it was producing thousands of lines when pulling images. I've just played with the settings on SecureCRT and found that if I disabled the "Line wrap" emulation mode, it stopped the issue from happening instantly. ![image](https://user-images.githubusercontent.com/35011199/193297358-c377b8b1-5755-4d75-94ee-7b77145cc554.png)
> 
> Apologies if this comment isn't of any relevance or significance but I found this issue when trying to troubleshoot the problem and wanted to share the fix that worked for me.

:-O I'm also using SecureCRT and the pull command has been horrible. Unchecking that box made a world of difference! Now I just have to figure out why the strange characters when it does it.
![2023-06-12_130825](https://github.com/docker/compose/assets/105168814/77f447c0-be51-4e24-9e2e-5545069968a1)

**Comment by Mariusio** [comment]
> > I'm at least still seeing this on Docker Compose version `v2.17.3` with docker version `23.0.5`
> 
> Same for me. I am using `Docker Desktop 4.19.0` on a M1 Macbook 16-inch, 2021 with Ventura 13.3.1.
> 
> Interestingly this bug only surfaced today after updating Docker Desktop from `4.17.x` to `4.19.0`, before the log output didn't jump and everything was smooth even with the same exact same environment and same `Dockerfile` and `docker-compose.yml`.

The issue stopped after I updated to `4.20.1 (110738)` today. Now the terminal output is behaving normally again,.

**Comment by williamdes** [comment]
@Mariusio This version does not exist here, is it the Docker version?

@ndeloof should all the recent reports comment elsewhere?

**Comment by ndeloof** [comment]
@williamdes this is Docker Desktop version
https://github.com/docker/compose/issues/10256 would indeed be a better place as an open issue

**Comment by stasadev** [comment]
```
$ docker-compose version
Docker Compose version 2.39.4
```

We also see broken output when pulling DDEV images if the terminal width is small. The progress line exceeds the terminal width, causing the output to break:

```
docker rmi ddev/ddev-dbserver-mariadb-10.11:v1.24.8 ddev/ddev-dbserver-mysql-8.0:v1.24.8
cat <<'EOF' | docker-compose -f - pull
services:
  ddev-ddev-dbserver-mariadb-10.11-v1.24.8:
    image: ddev/ddev-dbserver-mariadb-10.11:v1.24.8
  ddev-ddev-dbserver-mysql-8.0-v1.24.8:
    image: ddev/ddev-dbserver-mysql-8.0:v1.24.8
EOF
```

![Image](https://github.com/user-attachments/assets/5bb707f3-0a9a-4094-a4cd-c6e6d408f15d)

Here, `docker-compose` gives the progress bar `[ ]` too much space, making the line longer than the terminal width and breaking the output.

**Comment by simonschaufi** [comment]
Hi, sorry to notify you here but I'm a PHP developer and would like to know if somebody knows how docker is creating this really beautiful rotating (⠋⠼) and the building blocks progress bar? I would really appreciate a link to the file where it is generated.

I would like to build that for myself in PHP for the Symfony Console.

**Comment by ndeloof** [comment]
@simonschaufi this is all about ANSI sequences used to rewrite the current line.
See https://github.com/docker/compose/blob/main/cmd/display/spinner.go#L34

**Comment by williamdes** [comment]
And the progress display code might be in https://github.com/symfony/console/blob/7.3/Helper/ProgressIndicator.php

**Comment by simonschaufi** [comment]
Thank you very much! Now I just need to figure out how the Spinner is used. I can checkout the repo and inspect the code now at least.
